src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,116 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
+    public void growIfArrayIsFull() {
+        if (size == elements.length) {
+            int multiplalier = Math.round(elements.length * 1.5f);
+            System.arraycopy(elements, 0, elements = new Object[multiplalier], 0, size);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayIsFull();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            growIfArrayIsFull();
+            System.arraycopy(elements, index, elements, index + 1, Math.abs(index - size));
+            size++;
+            elements[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            return (T) elements[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            elements[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            final Object element = elements[index];
+            System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+            size--;
+            return (T) element;
```

Comment:
The index check logic is repeated in multiple methods. Consider extracting it into a private method to adhere to the DRY principle.